### ISSUE TITLE
Nested facts values

### ISSUE THREAD

**Issue by ctaity** [CacheControl/json-rules-engine#7]
Hi, this is a good library. But we need nested fact values, o use objects as facts, can be done?

Example:

var facts = { 
	test: {
		a: "a",
		b: "true"
	},
	displayMessage: true 
}

rule.setConditions({
  any: [{
    fact: 'test.b',
    operator: 'equal',
    value:  true 
  }]
})

**Comment by CacheControl** [comment]
Yes, this is supported.  You would do this by creating a fact function that extracts the appropriate field from the object, by referencing the rule params.  It would probably look something like:

```js
// define a fact's for extracting a specific field value from the 'test' object
let testField = (params, engine) => {
   return engine.factValue('test')  // this retrieves the top level "runtimeFacts" variable passed to engine.run(), below
   .then(testObj => {
     if (testObj[params.field]) {
        return testObj[params.field]   // params.field = 'b', so this effectively calls 'test.b'
     }
   })
}
engine.addFact('test-field', testField)  

let event = {
  type: 'testFieldTrigger'
}
let conditions = {
  any: [{
    fact: 'test-field',
    operator: 'equal',
    params: {
      field: 'b'
    },
    value: true
  }]
}
let rule = factories.rule({ conditions, event })
engine.addRule(rule)
let runtimeFacts = { 
  test: {
    a: "a",
    b: true   // CAREFUL!  In your example this was a "true" string; it should match the same data type as the condition.value; in this case, boolean.
  },
  displayMessage: true 
}
engine.run(runtimeFacts).then(results => console.log(results))
// fires:  [{ type: 'testFieldTrigger' }]
```

**Comment by ctaity** [comment]
Thanks, i see, i can access the root obejct (runtimeFact) from the engine. My idea is have a generic fact to access nested fields with lodash. Example

var nested_field = (params, engine) => {
   return engine.factValue().then(obj => {
     	return _.get(obj, params.field, null) ;
   })
}

and the rule

rule.setConditions({
  any: [{
    fact: 'nested_field',
    params: { field: "test.b" },
    operator: 'equal',
    value:  true 
  },{
    fact: 'nested_field',
    params: { field: "test.a" },
    operator: 'equal',
    value:  "testing" 
  }]
})

can be done?

**Comment by CacheControl** [comment]
@ctaity yep, you got it, that should work perfectly :)

**Comment by ctaity** [comment]
yes but engine.factValue() never resolve the promise.

**Comment by CacheControl** [comment]
Sounds like you might have a code bug, but I can't help identify it without more context.  

You might try getting a proof of concept going by building on an existing example, such as the [fact dependency](https://github.com/CacheControl/json-rules-engine/blob/master/examples/fact-dependency.js#L116) file; its the same concept as we're talking about above, and uses `factValue` in several places.

**Comment by slmmm** [comment]
I've written a couple of custom facts that I use to access properties in depth.
  **custom_fact_map_field.js**
`const _ = require('lodash');
module.exports = {
  name: 'map-field',
  fact: function (params, almanac) {
    let mapPathFields = _.split(params.fieldName, '.');
    let fieldMapPath = _.join(_.tail(mapPathFields), '.');
    return almanac.factValue('map-head', params)
      .then(function (response) {
        let fieldValue = _.get(response, fieldMapPath, null);
        if (fieldValue != null) {
          return fieldValue;
        } else {
          throw new Error('Undefined fact: ' + params.fieldName);
        }
      });
  }
};
`
  **custom_fact_map_head.js**
`const _ = require('lodash');
module.exports = {
  name: 'map-head',
  fact: function (params, almanac) {
    let mapPathFields = _.split(params.fieldName, '.');
    let mapPathRoot = _.head(mapPathFields);
    return almanac.factValue(mapPathRoot);
  }
};
`

Can be used in a rule like this
  **validate_address_parts.json**
`{
  "priority": 1,
  "event": {
    "type": "address-empty-fields",
    "params": {
      "action": "error",
      "message": "One or more of the address fields is empty",
    }
  },
  "conditions": {
    "any": [
      {
        "fact": "map-field",
        "params": {
          "fieldName": "address.street"
        },
        "operator": "equal",
        "value": ""
      },
      {
        "fact": "map-field",
        "params": {
          "fieldName": "address.zipcode"
        },
        "operator": "equal",
        "value": ""
      }
    ]
  }
}`

**Comment by ctaity** [comment]
Yes i do something very similar.

Is a way to extend operators?  I need to add some operators.

**Comment by ctaity** [comment]
In condition.js in function evaluate, can check is operator is a function???? or a way to extend condition.js with custom operators, example:

var ops = {
      inString : function(),
      not_contain: function(),
}

Condition.addOperatos(ops)

**Comment by CacheControl** [comment]
There's currently no support for custom operators, though that's an interesting feature request that we can look at adding.  Maybe something like:


```js
// register custom operator...
engine.addCustomOperator('my-custom-inString-operator', (factValue, value) => {
  if (typeof factValue !== 'string') return false; // because it's custom, you'll have to guard your own inputs
  return factValue.indexOf(value) // perform the comparison of the value the fact returns against the json "value" key.
})

// and usage would be the same as standard operators:
let condition = all: [{
      fact: 'my-fact',
      operator: 'my-custom-inString-operator',  // set custom operator name here
      value: 'comparison-string'
}]
engine.addRule({ condition, event })



// forgetting to call `addCustomOperator()` would throw an `Unknown operator` error when the engine encountered an unregistered operator.
```

Any thoughts?  Would this solve your need?

------

In the meantime, you might have an option to work around this.  You might create a fact that does your substring processing, and takes the data source fact as a parameter... something like:

```js
// this may work fine, or scale very poorly depending on your situation
{
    fact: 'inString',
    operator: 'equal',
    value: true,
    params: {
        fact: 'myFact',
        comparisonValue: 'substring-test'
      }
  }
```
So the 'inString' fact would call `factValue('myFact')`, and then compare the result with `substring-test`, and just return true or false based on whether it was a substring.  Then engine would then compare that the result equaled `true`.  

This is definitely a hack; if you like the custom operator solution above, I can probably get to it sometime in the next week.

**Comment by slmmm** [comment]
The ability to add custom operators would be great flexibility. 

Up to this point, I've been working around operations by creating custom facts.  For example, I needed to check for a substring within a string, so I created this custom fact and used in the following condition.

module.exports = {
  name: 'field-contains',
  fact: function (params, almanac) {
    return almanac.factValue(params.fieldName)
      .then(function (response) {
        if (response) {
          let found = -1;
             if (!params.isSensitive) {
               found = response.toLowerCase().search(params.value.toLowerCase());
             } else {
               found = response.search(params.value);
             }
           return (found >= 0);
         } else {
           return false;
         }
       })
       .catch(function () {
         return false;
       });
     }
}; 

// condition
{
  "fact": "field-contains",
  "params": {
    "fieldName": "lastName",
    "value": "Smith",
    "isSensitive": true
  },
  "operator": "equal",
  "value": true
}

**Comment by CacheControl** [comment]
@slmmm @ctaity This is great feedback!  I have some time later this week to take a crack at these suggestions; I'll tag you both on the PR(s) to get your thoughts.

**Comment by CacheControl** [comment]
@ctaity fyi, the two major features requests you had were added in #8 and #9.  Both are now available in version **1.2.0**.  Your final feature request, for event emissions after rule evaluations, is still on my radar.  I will leave this issue open until I've had a chance to add it.

**Comment by CacheControl** [comment]
@ctaity I've added your third and final request as part of #13 - individual rules now emit on 'success' and 'failure'.  Additionally, that request enabled a more elegant approach to the [rule chaining feature](https://github.com/CacheControl/json-rules-engine/blob/master/examples/07-rule-chaining.js), so thank you for that suggestion.

I'm closing this issue, as all enhancement requests have been fulfilled and published in **1.3.0**.

**Comment by ganeshcse2991** [comment]
Hi My Fact JSON is this :
{
status: "closed"
item:{
id: 1,
model: "something"
}
}

and My Rules is :
{
conditions: {
all: [
{
fact: "status",
operator: "equal",
value: "closed"
}
]
},
event: {
type: "sendEmail",
params: {
}
}
}

I am getting an error while running it. Engine is throwing error that it could not convert ObjectType to String. Since JSON object item is in the fact it trying to compare it as a string to my condition.

Please help where I am doing it wrong.
